### PR TITLE
Rename core notions (fixes #710)

### LINKED ISSUE #710: Rename core notions
Now that `cliquet` is in `kinto.core`, some naming is confusing.

In the core, resources are called collections, and objects are called records.

But, in Kinto, collections and records are instantiation of resources!

I suggest some renaming in kinto.core: 
- `collection` to `resource`
- `record` to `object`
- `collection_id` to `resource_name`  ~~(or `object_type` ?)~~
- `record_id` to `object_id`

That represents quite some work, but reducing confusion is always worth it.

### PR DESCRIPTION
**Pull request by leplatrem**
Fixes #710

- [x] Play with sed :)
- [x] Database migration
- [x] Repair stuff
- [x] Check openAPI (`-record`, `-collection`)
- [x] Look for leftovers
- [x] Deprecation paths where possible (decorator etc.)
- [ ] ~~List breaking changes~~   Make sure there's no breaking change, run tests from kinto-* plugins repo
- [ ] Add deprecation warning to kinto-redis storage backend

### CODE REVIEW

File: docs/api/1.x/_details-get-list.rst [hunk 1]
```diff
@@ -1,4 +1,4 @@
-A ``Total-Records`` response header indicates the total number of objects
+A ``Total-Objects`` response header indicates the total number of objects
```
Comment (Natim) [inline docs/api/1.x/_details-get-list.rst]: Objects or Items?

**Comment by leplatrem** [comment]
@glasserc what do you think of all this? Objects vs. Items, I thought we agreed on Objects. What about the "plural" things here and there?

**Comment by glasserc** [comment]
I spoke about this a little bit with @leplatrem on IRC. I have to say I like "items" better, but either one is an improvement on what we have now..

**Comment by glasserc** [comment]
Maybe even "resources", since it matches `resource_name`?

File: kinto/core/events.py [hunk 2]
```diff
@@ -233,15 +246,15 @@ def notify_resource_event(
...
     the previous one.
 
     :param resource_name: The name of the resource on which the event
         happened (taken from the request if not provided).
     :param resource_data: Information about the resource on which the
         event is being emitted. Usually contains information about how
-        to find this record in the hierarchy (for instance,
-        ``bucket_id`` and ``collection_id`` for a record). Taken from
+        to find this object in the hierarchy (for instance,
+        ``bucket_id`` and ``resource_name`` for a object). Taken from
```
Comment (glasserc) [inline kinto/core/events.py]: `an object`

File: kinto/core/cache/memcached.py [hunk 3]
```diff
@@ -86,7 +86,7 @@ def expire(self, key, ttl):
         if ttl == 0:
             self.delete(key)
         else:
-            # We can't use touch here because we need to update the TTL value in the record.
+            # We can't use touch here because we need to update the TTL value in the object.
```
Comment (glasserc) [inline kinto/core/cache/memcached.py]: This isn't really a Kinto record but a Memcached record. I could kind of go either way on this.

File: kinto/core/resource/__init__.py [hunk 4]
```diff
@@ -227,12 +228,12 @@ def timestamp(self):
...
             # If the instance is configured to be readonly, and if the
-            # collection is empty, the backend will try to bump the timestamp.
+            # resource is empty, the backend will try to bump the timestamp.
             # It fails if the configured db user has not write privileges.
             logger.exception(e)
             error_msg = (
-                "Collection timestamp cannot be written. "
-                "Records endpoint must be hit at least once from a "
+                "Resource timestamp cannot be written. "
+                "Objects endpoint must be hit at least once from a "
```
Comment (glasserc) [inline kinto/core/resource/__init__.py]: Should this be "plural endpoint", knowing what we know now?

File: kinto/core/resource/__init__.py [hunk 5]
```diff
@@ -279,25 +280,25 @@ def is_known_field(self, field):
...
 
         :raises: :exc:`~pyramid:pyramid.httpexceptions.HTTPNotModified` if
-            ``If-None-Match`` header is provided and collection not
+            ``If-None-Match`` header is provided and the resource not
             modified in the interim.
 
         :raises:
             :exc:`~pyramid:pyramid.httpexceptions.HTTPPreconditionFailed` if
-            ``If-Match`` header is provided and collection modified
+            ``If-Match`` header is provided and the resource modified
```
Comment (glasserc) [inline kinto/core/resource/__init__.py]: Should these be `and the items not modified in the interim`?

File: kinto/core/resource/__init__.py [hunk 6]
```diff
@@ -311,193 +312,195 @@ def collection_get(self):
...
 
-    def collection_delete(self):
-        """Model ``DELETE`` endpoint: delete multiple records.
+    def plural_delete(self):
+        """Model ``DELETE`` endpoint: delete multiple objects.
 
         :raises:
             :exc:`~pyramid:pyramid.httpexceptions.HTTPPreconditionFailed` if
-            ``If-Match`` header is provided and collection modified
+            ``If-Match`` header is provided and resource modified
```
Comment (glasserc) [inline kinto/core/resource/__init__.py]: Same question here about whether this should be `the elements` or something.

File: kinto/core/resource/model.py [hunk 7]
```diff
@@ -154,123 +157,159 @@ def delete_records(
...
-                record['index'] = idx
-                return record
+            def create_object(self, obj):
+                obj = super().create_object(obj)
+                idx = index.store(obj)
+                object['index'] = idx
+                return object
 
-        :param dict record: record to store
+        :param dict object: object to store
```
Comment (glasserc) [inline kinto/core/resource/model.py]: This should be `obj`

File: kinto/core/resource/model.py [hunk 8]
```diff
@@ -154,123 +157,159 @@ def delete_records(
...
+        :returns: the newly created object.
         :rtype: dict
         """
         parent_id = parent_id or self.parent_id
         return self.storage.create(
-            collection_id=self.collection_id,
+            resource_name=self.resource_name,
             parent_id=parent_id,
-            record=record,
+            object=obj,
```
Comment (glasserc) [inline kinto/core/resource/model.py]: Shouldn't this be `obj` too?

File: kinto/core/storage/__init__.py [hunk 9]
```diff
@@ -78,45 +79,45 @@ def flush(self, auth=None):
...
         """
         raise NotImplementedError
 
     def create(
         self,
-        collection_id,
+        resource_name,
         parent_id,
-        record,
+        object,
```
Comment (glasserc) [inline kinto/core/storage/__init__.py]: Shouldn't this be `obj`?

File: kinto/core/resource/model.py [hunk 10]
```diff
@@ -154,123 +157,159 @@ def delete_records(
...
-                subject = 'Record {} was changed'.format(record[self.id_field])
+            def update_object(self, obj, parent_id=None):
+                obj = super().update_object(obj, parent_id)
+                subject = 'Object {} was changed'.format(object[self.id_field])
                 send_email(subject)
-                return record
+                return object
 
-        :param dict record: record to store
+        :param dict object: object to store
```
Comment (glasserc) [inline kinto/core/resource/model.py]: `obj`

File: kinto/core/resource/model.py [hunk 11]
```diff
@@ -154,123 +157,159 @@ def delete_records(
...
-        record_id = record[self.id_field]
+        object_id = obj[self.id_field]
         return self.storage.update(
-            collection_id=self.collection_id,
+            resource_name=self.resource_name,
             parent_id=parent_id,
-            object_id=record_id,
-            record=record,
+            object_id=object_id,
+            object=obj,
```
Comment (glasserc) [inline kinto/core/resource/model.py]: `obj` ?

File: kinto/core/resource/model.py [hunk 12]
```diff
@@ -154,123 +157,159 @@ def delete_records(
...
+            def delete_object(self, obj):
+                deleted = super().delete_object(obj)
+                erase_media(obj)
                 deleted['media'] = 0
                 return deleted
 
-        :param dict record: the record to delete
-        :param dict record: record to store
+        :param dict object: the object to delete
+        :param dict object: object to store
```
Comment (glasserc) [inline kinto/core/resource/model.py]: Am I reading correctly? There are two `object` params? Also, `obj`

File: kinto/core/resource/model.py [hunk 13]
```diff
@@ -315,50 +354,50 @@ def delete_records(
...
-        perm_object_id = self.get_permission_object_id(record_id)
+        obj = super().get_object(object_id, parent_id)
+        perm_object_id = self.get_permission_object_id(object_id)
 
-        return self._annotate(record, perm_object_id)
+        return self._annotate(obj, perm_object_id)
 
-    def create_record(self, record, parent_id=None):
-        """Create record and set specified permissions.
+    def create_object(self, object, parent_id=None):
```
Comment (glasserc) [inline kinto/core/resource/model.py]: Shouldn't this be `obj`?

File: kinto/core/resource/model.py [hunk 14]
```diff
@@ -315,50 +354,50 @@ def delete_records(
...
+        perm_object_id = self.get_permission_object_id(object_id)
         self.permission.replace_object_permissions(perm_object_id, permissions)
         self._allow_write(perm_object_id)
 
-        return self._annotate(record, perm_object_id)
+        return self._annotate(obj, perm_object_id)
 
-    def update_record(self, record, parent_id=None):
-        """Update record and the specified permissions.
+    def update_object(self, object, parent_id=None):
```
Comment (glasserc) [inline kinto/core/resource/model.py]: `obj` ?

File: kinto/core/resource/schema.py [hunk 15]
```diff
@@ -400,27 +400,27 @@ class NotModifiedResponseSchema(colander.MappingSchema):
...
-class RecordResponseSchema(colander.MappingSchema):
+class ObjectResponseSchema(colander.MappingSchema):
     """Response schema used with sigle resource endpoints."""
 
     header = ResponseHeaderSchema()
 
     @colander.deferred
     def body(node, kwargs):
-        return kwargs.get("record")
+        return kwargs.get("object")
```
Comment (glasserc) [inline kinto/core/resource/schema.py]: I'm not sure I really understand what these `kwargs` are and whether I should expect them to have an `object` or a `record`.

File: kinto/core/scripts.py [hunk 16]
```diff
@@ -47,36 +47,36 @@ def delete_collection(env, bucket_id, collection_id):
...
-        registry.storage.get(collection_id="collection", parent_id=bucket, object_id=collection_id)
-    except storage_exceptions.RecordNotFoundError:
+        registry.storage.get(resource_name="collection", parent_id=bucket, object_id=collection_id)
+    except storage_exceptions.ObjectNotFoundError:
         logger.error("Collection '{}' does not exist.".format(collection))
         return 33
 
     deleted = registry.storage.delete_all(
-        collection_id="record", parent_id=collection, with_deleted=False
+        resource_name="object", parent_id=collection, with_deleted=False
```
Comment (glasserc) [inline kinto/core/scripts.py]: This should remain `record`

File: kinto/core/scripts.py [hunk 17]
```diff
@@ -47,36 +47,36 @@ def delete_collection(env, bucket_id, collection_id):
...
     )
     if len(deleted) == 0:
-        logger.info("No records found for '{}'.".format(collection))
+        logger.info("No objects found for '{}'.".format(collection))
     else:
-        logger.info("{} record(s) were deleted.".format(len(deleted)))
+        logger.info("{} object(s) were deleted.".format(len(deleted)))
```
Comment (glasserc) [inline kinto/core/scripts.py]: Same with these.

File: kinto/core/scripts.py [hunk 18]
```diff
@@ -47,36 +47,36 @@ def delete_collection(env, bucket_id, collection_id):
...
 
     registry.storage.delete(
-        collection_id="collection", parent_id=bucket, object_id=collection_id, with_deleted=False
+        resource_name="collection", parent_id=bucket, object_id=collection_id, with_deleted=False
     )
     logger.info("'{}' collection object was deleted.".format(collection))
 
-    record = "/buckets/{bucket_id}" "/collections/{collection_id}" "/records/{record_id}"
+    obj = "/buckets/{bucket_id}" "/collections/{collection_id}" "/objects/{object_id}"
```
Comment (glasserc) [inline kinto/core/scripts.py]: These too.

File: kinto/core/storage/__init__.py [hunk 19]
```diff
@@ -10,7 +11,7 @@
 
 
 class Missing:
-    """Dummy value to represent a value that is completely absent from a record.
+    """Dummy value to represent a value that is completely absent from a object.
```
Comment (glasserc) [inline kinto/core/storage/__init__.py]: `an object`

File: kinto/core/storage/__init__.py [hunk 20]
```diff
@@ -356,7 +362,7 @@ def ping(request):
                     backend.delete_all(**storage_kw)
                     backend.purge_deleted(**storage_kw)  # Kinto/kinto#985
                 else:
-                    backend.create(record=_HEARTBEAT_RECORD, **storage_kw)
+                    backend.create(object=_HEARTBEAT_OBJECT, **storage_kw)
```
Comment (glasserc) [inline kinto/core/storage/__init__.py]: Should this be `obj`?

File: kinto/core/storage/exceptions.py [hunk 21]
```diff
@@ -32,12 +40,12 @@ class UnicityError(IntegrityError):
...
+    object violates the unicity constraints defined by the resource.
 
     """
 
-    def __init__(self, field, record, *args, **kwargs):
+    def __init__(self, field, obj, *args, **kwargs):
         self.field = field
-        self.record = record
-        self.msg = "{} is not unique: {}".format(field, record)
+        self.object = object
```
Comment (glasserc) [inline kinto/core/storage/exceptions.py]: This should be `obj`.

File: kinto/core/storage/testing.py [hunk 22]
```diff
@@ -24,8 +24,8 @@ def setUp(self):
         self.modified_field = "last_modified"
         self.client_error_patcher = None
 
-        self.record = {"foo": "bar"}
-        self.storage_kw = {"collection_id": "test", "parent_id": "1234", "auth": "Basic bWF0OjI="}
+        self.object = {"foo": "bar"}
```
Comment (glasserc) [inline kinto/core/storage/testing.py]: More generally, I don't love `self.object`, especially if we're using `obj` to avoid the builtin `object`.

**Comment by leplatrem** [comment]
This is ready for merge :)

File: CHANGELOG.rst [hunk 23]
```diff
@@ -18,11 +22,9 @@ This document describes changes between each past release.
 - Remove regexp py36 warnings. (#1907)
 - Changed psycopg2 dependency for psycopg2-binary. (#1905)
 - Still supports jsonschema 2.6 before 3.0 is released as a production release. (#1923)
+- Renamed core notions (ie. record and collection) (#710)
```
[the remark this answers is not in the document]
Comment (leplatrem) [reply CHANGELOG.rst]: Well, in theory I left the old methods with deprecation warnings

File: tests/core/resource/test_pagination.py [hunk 24]
```diff
@@ -252,18 +252,18 @@ def test_paginated_delete_second_to_last_gets_next_header(self):
...
-        self.assertRaises(HTTPBadRequest, self.resource.collection_delete)
+        self.assertRaises(HTTPBadRequest, self.resource.plural_delete)
 
 
 class BuildPaginationTokenTest(BaseTest):
     def setUp(self):
         super().setUp()
         self.patch_known_field.start()
-        self.record = {
+        self.object = {
```
[the remark this answers is not in the document]
Comment (leplatrem) [reply tests/core/resource/test_pagination.py]: I'm not shocked by `self.object`, but apparently you are, so I'll change it...

File: CHANGELOG.rst [hunk 25]
```diff
@@ -10,6 +10,8 @@ This document describes changes between each past release.
 
 - Remove Python 3.5 support and upgrade to Python 3.6. (#1886)
 - Remove `record` from UnicityError class (#1919). This enabled us to fix #1545.
+- Storage backend API has changed, notions of collection and records were replaced
+  by the generic terms *resource* and *object*.
```
Comment (glasserc) [inline CHANGELOG.rst]: It would (still...) nice to add a note here mentioning that this could break you if you implemented a resource subclass.

**Comment by leplatrem** [comment]
> hopefully you didn't rebase any changes into an earlier file.

Nope I didn't. I force-pushed because I amended the merge commit from master.

Thanks for your reviews! It's ready to land! Since this touches many files, I'm tempted to squash the commits before merging...

**Pull request merged**